File: android/core/src/main/kotlin/org/hdhropen/kit/networking/SyncPlayClient.kt
package org.hdhropen.kit.networking

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.hdhropen.kit.models.*
import org.hdhropen.kit.utilities.Log

class SyncPlayClient(
    private val httpClient: OkHttpClient,
    private val json: Json,
    private val coroutineScope: CoroutineScope
) {
    private var webSocket: WebSocket? = null
    private var pingJob: Job? = null
    private var progressJob: Job? = null

    private val _room = MutableStateFlow<SyncPlayRoom?>(null)
    val room: StateFlow<SyncPlayRoom?> = _room.asStateFlow()

    private val _participants = MutableStateFlow<List<SyncPlayParticipant>>(emptyList())
    val participants: StateFlow<List<SyncPlayParticipant>> = _participants.asStateFlow()

    private val _sessionId = MutableStateFlow<String?>(null)
    val sessionId: StateFlow<String?> = _sessionId.asStateFlow()

    private val _isHost = MutableStateFlow(false)
    val isHost: StateFlow<Boolean> = _isHost.asStateFlow()

    private val _pingMs = MutableStateFlow(0.0)
    val pingMs: StateFlow<Double> = _pingMs.asStateFlow()

    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()

    var onRemotePlay: ((position: Double, rate: Double) -> Unit)? = null
    var onRemotePause: ((position: Double) -> Unit)? = null
    var onRemoteSeek: ((position: Double) -> Unit)? = null
    var onRemoteContentChange: ((content: SyncPlayContent) -> Unit)? = null
    var getCurrentPosition: (() -> Double)? = null
    var isPlayerReady: (() -> Boolean)? = null

    fun connect(wsUrl: String) {
        disconnect()

        val request = Request.Builder()
            .url(wsUrl)
            .build()

        webSocket = httpClient.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                _isConnected.value = true
                startPingAndProgressJobs()
                Log.network.info("SyncPlay WebSocket connected")
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                try {
                    val msg = json.decodeFromString<SyncPlayMessage>(text)
                    handleMessage(msg)
                } catch (e: Exception) {
                    Log.network.error("SyncPlay JSON decode error: ${e.localizedMessage}")
                }
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                webSocket.close(1000, null)
                handleDisconnected()
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                Log.network.error("SyncPlay WebSocket failure: ${t.localizedMessage}")
                handleDisconnected()
            }
        })
    }

    private fun handleMessage(msg: SyncPlayMessage) {
        when (msg.type) {
            "room_state" -> {
                msg.room?.let { r ->
                    _room.value = r
                    _participants.value = r.participants
                    msg.yourSessionId?.let { sid ->
                        _sessionId.value = sid
                        _isHost.value = (r.hostSessionId == sid)
                    }
                }
            }
            "participant_joined" -> {
                msg.participant?.let { p ->
                    val current = _participants.value.toMutableList()
                    if (current.none { it.sessionId == p.sessionId }) {
                        current.add(p)
                        _participants.value = current
                    }
                }
            }
            "participant_left" -> {
                val leftId = msg.sessionId
                if (leftId != null) {
                    val current = _participants.value.filter { it.sessionId != leftId }
                    _participants.value = current
                    msg.newHostSessionId?.let { newHostId ->
                        _isHost.value = (_sessionId.value == newHostId)
                        _participants.value = current.map {
                            if (it.sessionId == newHostId) it.copy(isHost = true) else it
                        }
                    }
                }
            }
            "participant_updated" -> {
                msg.participant?.let { updated ->
                    _participants.value = _participants.value.map {
                        if (it.sessionId == updated.sessionId) updated else it
                    }
                }
            }
            "playback_update" -> {
                val action = msg.action
                val pos = msg.position ?: 0.0
                val rate = msg.playbackRate ?: 1.0
                val isPlaying = msg.isPlaying ?: false
                val serverTime = msg.serverTime ?: 0.0

                _room.value?.let { currentRoom ->
                    _room.value = currentRoom.copy(
                        playbackState = SyncPlayPlaybackState(
                            isPlaying = isPlaying,
                            position = pos,
                            playbackRate = rate,
                            updatedAt = serverTime
                        )
                    )
                }

                // If triggered by a remote peer, apply to local player
                if (msg.triggeredBy != _sessionId.value) {
                    when (action) {
                        "play" -> onRemotePlay?.invoke(pos, rate)
                        "pause" -> onRemotePause?.invoke(pos)
                        "seek" -> onRemoteSeek?.invoke(pos)
                    }
                }
            }
            "content_changed" -> {
                msg.content?.let { c ->
                    onRemoteContentChange?.invoke(c)
                }
                msg.room?.let { r ->
                    _room.value = r
                    _participants.value = r.participants
                }
            }
            "host_changed" -> {
                msg.newHostSessionId?.let { newHostId ->
                    _isHost.value = (_sessionId.value == newHostId)
                    _participants.value = _participants.value.map {
                        it.copy(isHost = it.sessionId == newHostId)
                    }
                }
            }
            "pong" -> {
                val clientTime = msg.clientTime ?: 0.0
                val now = System.currentTimeMillis().toDouble()
                val rtt = now - clientTime
                _pingMs.value = maxOf(1.0, rtt)
            }
        }
    }

    private fun startPingAndProgressJobs() {
        pingJob?.cancel()
        pingJob = coroutineScope.launch(Dispatchers.IO) {
            while (isActive && _isConnected.value) {
                delay(4000)
                val pingObj = mapOf(
                    "type" to "ping",
                    "client_time" to System.currentTimeMillis().toDouble()
                )
                sendMessage(pingObj)
            }
        }

        progressJob?.cancel()
        progressJob = coroutineScope.launch(Dispatchers.IO) {
            while (isActive && _isConnected.value) {
                delay(2000)
                val pos = getCurrentPosition?.invoke() ?: 0.0
                val ready = isPlayerReady?.invoke() ?: true
                val progressObj = mapOf(
                    "type" to "progress",
                    "position" to pos,
                    "is_ready" to ready,
                    "ping_ms" to _pingMs.value
                )
                sendMessage(progressObj)
            }
        }
    }

    fun sendPlay(position: Double, playbackRate: Double = 1.0) {
        sendMessage(mapOf(
            "type" to "play",
            "position" to position,
            "playback_rate" to playbackRate
        ))
    }

    fun sendPause(position: Double) {
        sendMessage(mapOf(
            "type" to "pause",
            "position" to position
        ))
    }

    fun sendSeek(position: Double) {
        sendMessage(mapOf(
            "type" to "seek",
            "position" to position
        ))
    }

    fun changeContent(content: SyncPlayContent) {
        val jsonContent = json.encodeToString(SyncPlayContent.serializer(), content)
        sendMessage(mapOf(
            "type" to "change_content",
            "content" to jsonContent
        ))
    }

    fun transferHost(targetSessionId: String) {
        sendMessage(mapOf(
            "type" to "transfer_host",
            "target_session_id" to targetSessionId
        ))
    }

    private fun sendMessage(map: Map<String, Any>) {
        try {
            val str = map.entries.joinToString(prefix = "{", postfix = "}") { (k, v) ->
                val vStr = when (v) {
                    is Number, is Boolean -> v.toString()
                    is String -> if (v.startsWith("{")) v else "\"$v\""
                    else -> "\"$v\""
                }
                "\"$k\":$vStr"
            }
            webSocket?.send(str)
        } catch (e: Exception) {
            Log.network.error("Error sending SyncPlay message: ${e.localizedMessage}")
        }
    }

    private fun handleDisconnected() {
        _isConnected.value = false
        _room.value = null
        _participants.value = emptyList()
        _sessionId.value = null
        _isHost.value = false
        pingJob?.cancel()
        progressJob?.cancel()
    }

    fun disconnect() {
        try {
            webSocket?.send("""{"type":"leave"}""")
            webSocket?.close(1000, "User left")
        } catch (_: Exception) {}
        webSocket = null
        handleDisconnected()
    }
}

File: apple/HDHROpenKit/Sources/HDHROpenKit/Networking/SyncPlayClient.swift
import Combine
import Foundation

@MainActor
public final class SyncPlayClient: ObservableObject {
    @Published public private(set) var room: SyncPlayRoom?
    @Published public private(set) var participants: [SyncPlayParticipant] = []
    @Published public private(set) var sessionId: String?
    @Published public private(set) var isHost = false
    @Published public private(set) var pingMs = 0.0
    @Published public private(set) var isConnected = false

    public var onRemotePlay: ((Double, Double) -> Void)?
    public var onRemotePause: ((Double) -> Void)?
    public var onRemoteSeek: ((Double) -> Void)?
    public var onRemoteContentChange: ((SyncPlayContent) -> Void)?
    public var getCurrentPosition: (() -> Double)?
    public var isPlayerReady: (() -> Bool)?

    private var webSocketTask: URLSessionWebSocketTask?
    private var pingTimer: Task<Void, Never>?
    private var progressTimer: Task<Void, Never>?
    private let jsonDecoder = JSONDecoder()
    private let jsonEncoder = JSONEncoder()

    public init() {}

    public func connect(url: URL) {
        disconnect()

        let session = URLSession(configuration: .default)
        let task = session.webSocketTask(with: url)
        webSocketTask = task
        isConnected = true
        task.resume()

        receiveMessage()
        startHeartbeats()
        Log.network.info("SyncPlay WebSocket connected to \(url.absoluteString)")
    }

    public func disconnect() {
        sendJson(["type": "leave"])
        webSocketTask?.cancel(with: .normalClosure, reason: nil)
        webSocketTask = nil
        handleDisconnected()
    }

    private func handleDisconnected() {
        isConnected = false
        room = nil
        participants = []
        sessionId = nil
        isHost = false
        pingTimer?.cancel()
        pingTimer = nil
        progressTimer?.cancel()
        progressTimer = nil
    }

    private func receiveMessage() {
        guard let task = webSocketTask else { return }
        task.receive { [weak self] result in
            Task { @MainActor in
                guard let self, self.isConnected else { return }
                switch result {
                case let .success(message):
                    switch message {
                    case let .string(text):
                        self.handleRawMessage(text)
                    case let .data(data):
                        if let text = String(data: data, encoding: .utf8) {
                            self.handleRawMessage(text)
                        }
                    @unknown default:
                        break
                    }
                    self.receiveMessage()
                case let .failure(error):
                    Log.network.error("SyncPlay WebSocket receive error: \(error.localizedDescription)")
                    self.handleDisconnected()
                }
            }
        }
    }

    public func handleRawMessage(_ text: String) {
        guard let data = text.data(using: .utf8) else { return }
        do {
            let msg = try jsonDecoder.decode(SyncPlayMessage.self, from: data)
            handleParsedMessage(msg)
        } catch {
            Log.network.error("SyncPlay JSON decode error: \(error.localizedDescription)")
        }
    }

    private func handleParsedMessage(_ msg: SyncPlayMessage) {
        switch msg.type {
        case "room_state":
            if let r = msg.room {
                room = r
                participants = r.participants
                if let sid = msg.yourSessionId {
                    sessionId = sid
                    isHost = (r.hostSessionId == sid)
                }
            }
        case "participant_joined":
            if let p = msg.participant {
                if !participants.contains(where: { $0.sessionId == p.sessionId }) {
                    participants.append(p)
                }
            }
        case "participant_left":
            if let leftId = msg.sessionId {
                participants.removeAll(where: { $0.sessionId == leftId })
                if let newHostId = msg.newHostSessionId {
                    isHost = (sessionId == newHostId)
                    participants = participants.map {
                        SyncPlayParticipant(
                            sessionId: $0.sessionId,
                            userName: $0.userName,
                            isHost: $0.sessionId == newHostId,
                            isReady: $0.isReady,
                            position: $0.position,
                            pingMs: $0.pingMs
                        )
                    }
                }
            }
        case "participant_updated":
            if let updated = msg.participant {
                participants = participants.map {
                    $0.sessionId == updated.sessionId ? updated : $0
                }
            }
        case "playback_update":
            let action = msg.action
            let pos = msg.position ?? 0.0
            let rate = msg.playbackRate ?? 1.0
            let isPlaying = msg.isPlaying ?? false
            let serverTime = msg.serverTime

            if let currentRoom = room {
                let newState = SyncPlayPlaybackState(
                    isPlaying: isPlaying,
                    position: pos,
                    playbackRate: rate,
                    updatedAt: serverTime
                )
                room = SyncPlayRoom(
                    roomCode: currentRoom.roomCode,
                    hostSessionId: currentRoom.hostSessionId,
                    createdAt: currentRoom.createdAt,
                    playbackState: newState,
                    currentContent: currentRoom.currentContent,
                    participants: participants
                )
            }

            if msg.triggeredBy != sessionId {
                switch action {
                case "play":
                    onRemotePlay?(pos, rate)
                case "pause":
                    onRemotePause?(pos)
                case "seek":
                    onRemoteSeek?(pos)
                default:
                    break
                }
            }
        case "content_changed":
            if let content = msg.content {
                onRemoteContentChange?(content)
            }
            if let r = msg.room {
                room = r
                participants = r.participants
            }
        case "host_changed":
            if let newHostId = msg.newHostSessionId {
                isHost = (sessionId == newHostId)
                participants = participants.map {
                    SyncPlayParticipant(
                        sessionId: $0.sessionId,
                        userName: $0.userName,
                        isHost: $0.sessionId == newHostId,
                        isReady: $0.isReady,
                        position: $0.position,
                        pingMs: $0.pingMs
                    )
                }
            }
        case "pong":
            if let clientTime = msg.clientTime {
                let now = Date().timeIntervalSince1970 * 1000.0
                let rtt = max(1.0, now - clientTime)
                pingMs = rtt
            }
        default:
            break
        }
    }

    private func startHeartbeats() {
        pingTimer?.cancel()
        pingTimer = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 4_000_000_000)
                guard let self, isConnected else { break }
                let now = Date().timeIntervalSince1970 * 1000.0
                sendJson(["type": "ping", "client_time": now])
            }
        }

        progressTimer?.cancel()
        progressTimer = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 2_000_000_000)
                guard let self, isConnected else { break }
                let pos = getCurrentPosition?() ?? 0.0
                let ready = isPlayerReady?() ?? true
                sendJson([
                    "type": "progress",
                    "position": pos,
                    "is_ready": ready,
                    "ping_ms": pingMs
                ])
            }
        }
    }

    public func sendPlay(position: Double, playbackRate: Double = 1.0) {
        sendJson([
            "type": "play",
            "position": position,
            "playback_rate": playbackRate
        ])
    }

    public func sendPause(position: Double) {
        sendJson([
            "type": "pause",
            "position": position
        ])
    }

    public func sendSeek(position: Double) {
        sendJson([
            "type": "seek",
            "position": position
        ])
    }

    public func changeContent(_ content: SyncPlayContent) {
        do {
            let contentData = try jsonEncoder.encode(content)
            if let contentStr = String(data: contentData, encoding: .utf8) {
                sendJson([
                    "type": "change_content",
                    "content": contentStr
                ])
            }
        } catch {
            Log.network.error("Error encoding content for SyncPlay: \(error.localizedDescription)")
        }
    }

    public func transferHost(targetSessionId: String) {
        sendJson([
            "type": "transfer_host",
            "target_session_id": targetSessionId
        ])
    }

    private func sendJson(_ dict: [String: Any]) {
        guard let task = webSocketTask else { return }
        do {
            let data = try JSONSerialization.data(withJSONObject: dict, options: [])
            if let str = String(data: data, encoding: .utf8) {
                task.send(.string(str)) { error in
                    if let err = error {
                        Log.network.error("SyncPlay send error: \(err.localizedDescription)")
                    }
                }
            }
        } catch {
            Log.network.error("SyncPlay JSON serialize error: \(error.localizedDescription)")
        }
    }
}
